### PR TITLE
Add an Open in browser button on error panel

### PR DESCRIPTION
**Pull request by YonghaoDeng**
#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR

With the addition of the open_in_browser button design, the The user can watch the video in case the new pipe itself is faulty.

#### Before/After Screenshots/Screen Record

<table>
  <tr>
    <th>Before</th>
    <th>After</th>
  </tr>
  <tr>
    <td>Before</td>
    <td>After</td>
  </tr>
</table>

#### Fixes the following issue(s)

- Fixes  https://github.com/TeamNewPipe/NewPipe/issues/8800

#### APK testing

The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

File: app/src/main/java/org/schabi/newpipe/error/ErrorInfo.kt [hunk 1]
```diff
@@ -82,10 +82,8 @@ class ErrorInfo(
 
         fun throwableListToStringList(throwableList: List<Throwable>) =
             throwableList.map { it.stackTraceToString() }.toTypedArray()
-
         private fun getInfoServiceName(info: Info?) =
             if (info == null) SERVICE_NONE else ServiceHelper.getNameOfServiceById(info.serviceId)
-
```
Comment (AudricV) [inline app/src/main/java/org/schabi/newpipe/error/ErrorInfo.kt:88]: Remove these unneeded changes.

File: app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt [hunk 2]
```diff
@@ -69,19 +73,17 @@ class ErrorPanelHelper(
         errorServiceExplanationTextView.isVisible = false
         errorActionButton.isVisible = false
         errorRetryButton.isVisible = false
+        errorOpenbrowserButton.isVisible = false
     }
 
     fun showError(errorInfo: ErrorInfo) {
-
```
Comment (AudricV) [inline app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt:75]: Same as above.

File: app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt [hunk 3]
```diff
@@ -69,19 +73,17 @@ class ErrorPanelHelper(
...
         if (errorInfo.throwable != null && errorInfo.throwable!!.isInterruptedCaused) {
             if (DEBUG) {
                 Log.w(TAG, "onError() isInterruptedCaused! = [$errorInfo.throwable]")
             }
             return
         }
 
-        ensureDefaultVisibility()
-
```
Comment (AudricV) [inline app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt:84]: Could you explain why you removed this?

File: app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt [hunk 4]
```diff
@@ -127,6 +130,12 @@ class ErrorPanelHelper(
             ) {
                 // show retry button only for content which is not unavailable or unsupported
                 errorRetryButton.isVisible = true
+                errorOpenbrowserButton.isVisible = true
+                errorOpenbrowserButton.setOnClickListener {
+                    val uriUrl = Uri.parse(errorInfo.request)
+                    val intent = Intent(Intent.ACTION_VIEW, uriUrl)
+                    context.startActivity(intent)
```
Comment (AudricV) [inline app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt:137]: Use [`ShareUtils.openUrlInBrowser`](https://github.com/TeamNewPipe/NewPipe/blob/2a98cca80100d87562f0c9ff1fadeac3838f6ff6/app/src/main/java/org/schabi/newpipe/util/external_communication/ShareUtils.java#L125) here instead.

File: app/src/main/res/layout/error_panel.xml [hunk 5]
```diff
@@ -72,4 +72,19 @@
         android:visibility="gone"
         tools:visibility="visible" />
 
+    <Button
+        android:id="@+id/error_openinbrowser"
```
Comment (AudricV) [inline app/src/main/res/layout/error_panel.xml:76]:
```suggestion
        android:id="@+id/error_open_in_browser"
```

File: app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt [hunk 6]
```diff
@@ -52,6 +54,8 @@ class ErrorPanelHelper(
         errorPanelRoot.findViewById(R.id.error_action_button)
     private val errorRetryButton: Button =
         errorPanelRoot.findViewById(R.id.error_retry_button)
+    private val errorOpenbrowserButton: Button =
+        errorPanelRoot.findViewById(R.id.error_openinbrowser)
```
Comment (AudricV) [inline app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt:58]:
```suggestion
    private val errorOpenInBrowserButton: Button =
        errorPanelRoot.findViewById(R.id.error_open_in_browser)
```

File: app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt [hunk 7]
```diff
@@ -69,19 +73,17 @@ class ErrorPanelHelper(
         errorServiceExplanationTextView.isVisible = false
         errorActionButton.isVisible = false
         errorRetryButton.isVisible = false
+        errorOpenbrowserButton.isVisible = false
```
Comment (AudricV) [inline app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt:76]:
```suggestion
        errorOpenInBrowserButton.isVisible = false
```

File: app/src/main/res/layout/error_panel.xml [hunk 8]
```diff
@@ -72,4 +72,19 @@
...
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="4dp"
+        android:layout_marginBottom="8dp"
```
Comment (AudricV) [inline app/src/main/res/layout/error_panel.xml:80]: As you can notice on the screenshot you provided, the margin between the `Open in browser` and `Retry` buttons is not the same as the one between the `Report` and `Retry` buttons. Could you change this?

File: app/src/main/res/layout/error_panel.xml [hunk 9]
```diff
@@ -72,4 +72,19 @@
...
+        android:text="@string/error_openinbrowser"
```
Comment (AudricV) [inline app/src/main/res/layout/error_panel.xml:81]: There is already a string for `Open in browser`: https://github.com/TeamNewPipe/NewPipe/blob/2a98cca80100d87562f0c9ff1fadeac3838f6ff6/app/src/main/res/values/strings.xml#L10

```suggestion
        android:text="@string/open_in_browser"
```

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -251,6 +251,7 @@
     <string name="error_report_open_github_notice">Please check whether an issue discussing your crash already exists. When creating duplicate tickets, you take time from us which we could spend with fixing the actual bug.</string>
     <string name="error_snackbar_message">Sorry, something went wrong.</string>
     <string name="error_snackbar_action">Report</string>
+    <string name="error_openinbrowser">Open in Browser</string>
```
Comment (AudricV) [inline app/src/main/res/values/strings.xml:254]: See above.

**Review by AudricV — CHANGES_REQUESTED** [review]
Thank you for this improvement! Here are changes that you need to apply in order to have this pull request merged.

File: app/src/main/java/org/schabi/newpipe/error/ErrorInfo.kt (hunk 1 above)
Comment (YonghaoDeng) [reply app/src/main/java/org/schabi/newpipe/error/ErrorInfo.kt:88]: I didn't change this line of code, it was already there, are you sure I want to remove it?

File: app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt (hunk 2 above)
Comment (YonghaoDeng) [reply app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt:75]: I didn't change this line of code either, it was already there, are you sure I want to remove it? Also if I delete it, the file will error out

File: app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt (hunk 3 above)
Comment (YonghaoDeng) [reply app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt:84]: It's my fault, I shouldn't have deleted it

File: app/src/main/java/org/schabi/newpipe/error/ErrorInfo.kt (hunk 1 above)
Comment (AudricV) [reply app/src/main/java/org/schabi/newpipe/error/ErrorInfo.kt:88]: In the case you didn't understand, I meant don't remove these lines with your changes in this PR.

File: app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt (hunk 2 above)
Comment (AudricV) [reply app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt:75]: See my last comment above.

File: app/src/main/java/org/schabi/newpipe/error/ErrorInfo.kt (hunk 1 above)
Comment (YonghaoDeng) [reply app/src/main/java/org/schabi/newpipe/error/ErrorInfo.kt]: Oh! make sense!

Comment (YonghaoDeng) [reply app/src/main/java/org/schabi/newpipe/error/ErrorInfo.kt]: I have modified!

**Comment by opusforlife2** [comment]
What will the button do, BTW? Will it directly open the default browser selected in the OS? Or show a list of apps that can handle the link type?

**Comment by YonghaoDeng** [comment]
> it will directly open the default browser selected in the OS

**Comment by opusforlife2** [comment]
I think it would be better to show the list of apps. The user may want to open it in an app other than the default browser, like a media player.

**Comment by YonghaoDeng** [comment]
> But I think the video needs to be viewed online at the YouTube website and the requirement is to create an open in browser button, so I think it should be redirected to the browser, don't you think so?

**Comment by opusforlife2** [comment]
It's perfectly fine to open it in the default browser. Nothing wrong with that behaviour.

I'm just providing a suggestion for extra functionality. If a user wants to open it in a different browser, or in a media app, they should be able to. They can always select "Always open" so that their selected choice becomes the default way to open.

**Comment by YonghaoDeng** [comment]
> show the list of apps

Hmmmmm I see what you mean but I'm afraid I won't be able to do well with this extra feature. As I am an Anu student, this is an example of PR practice in one of my classes and the deadline for this assignment is this weekend

**Comment by opusforlife2** [comment]
No problems at all. 👍

File: settings.gradle [hunk 11]
```diff
@@ -4,8 +4,8 @@ include ':app'
 // We assume, that NewPipe and NewPipe Extractor have the same parent directory.
 // If this is not the case, please change the path in includeBuild().
 
-//includeBuild('../NewPipeExtractor') {
```
Comment (Stypox) [inline settings.gradle:7]: Do not push these changes, they should be local-only

File: app/src/main/res/layout/error_panel.xml [hunk 12]
```diff
@@ -72,4 +72,19 @@
         android:visibility="gone"
         tools:visibility="visible" />
 
+    <Button
+        android:id="@+id/error_open_in_browser"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="-4dp"
```
Comment (Stypox) [inline app/src/main/res/layout/error_panel.xml:79]: Remove `layout_marginBottom` from `error_retry_button` and remove this line: negative margins are not a good idea, and also at the moment the buttons are not equally spaced out.

File: app/src/main/res/layout/error_panel.xml [hunk 13]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools"
-    android:layout_width="wrap_content"
+    android:layout_width="match_parent"
```
Comment (Stypox) [inline app/src/main/res/layout/error_panel.xml:4]: Why this change?

**Review by Stypox — CHANGES_REQUESTED** [review]
Thank you! Looks almost good

File: app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt [hunk 14]
```diff
@@ -99,6 +103,7 @@ class ErrorPanelHelper(
             }
 
             errorRetryButton.isVisible = true
+            errorOpenInBrowserButton.isVisible = true
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt:106]: Shouldn't the on click listener be set here, too? I suggest that you create a `showAndSetOpenInBrowserButtonAction` function which sets the button to visible and sets the on click listener. Then you can call that function both here and 30 lines below here ;-)

**Comment by YonghaoDeng** [comment]
I added the function as you said @Stypox

File: app/src/main/res/layout/error_panel.xml [hunk 15]
```diff
@@ -62,7 +62,7 @@
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"
         android:layout_marginTop="4dp"
-        android:layout_marginBottom="8dp"
+
```
Comment (Stypox) [inline app/src/main/res/layout/error_panel.xml:65]: Remove this blank line
```suggestion
```

File: app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt [hunk 16]
```diff
@@ -127,6 +132,7 @@ class ErrorPanelHelper(
             ) {
                 // show retry button only for content which is not unavailable or unsupported
                 errorRetryButton.isVisible = true
+                showAndSetOpenInBrowserButtonAction(errorInfo)
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt:135]: I think it's better if you put this line outside the `if (errorInfo.throwable !is ContentNotAvailableException &&
                errorInfo.throwable !is ContentNotSupportedException
            )`, since it's ok to allow the user to go to the browser even when the content is unavailable or unsupported

**Review by Stypox — CHANGES_REQUESTED** [review]
Just a few more little changes

**Comment by YonghaoDeng** [comment]
OK，I modified the code as you said @Stypox

**Review by Stypox — APPROVED** [review]
Looks good to me. Thank you!
